[Downey]Step 2 - DownCasting, Set, Access Control

dahun-lee-daji opened the pull request:
# 작업 목록 
## 음료 클래스
- [x] 음료 클래스 3단계 이상 구현
- [x] 상위 클래스에 없는 하나 이상의 속성
- [x] 속성에 대한 추가 인터페이스
- [x] 추가 인터페이스
- [x] testCode

## 자판기
- [x] 원하는 금액만큼 올리는 메소드
- [x] 재고 추가 메소드
- [x] 구매가능 목록 리턴 메소드
- [x] 음료수 구매 메소드
- [x] 잔액 확인 메소드
- [x] Dictionary Type 재고 return method
- [x] 유통기한 지난 재고 return method
- [x] 따뜻한 음료 return method
- [x] 구매 이력 return method
- [x] testCode 

# 학습 키워드
DownCasting, Set, Access Control

# 고민과 해결
## Access Control
이전에, 상위 모듈에서 하위 모듈의 property에 접근하면 안된다는 가르침이 있기에, 이를 따르기 위해 class의 property를 private으로 선언하여 사용했습니다. 하지만 상위 모듈에서 Drink.name이 필요한 경우가 다수 발생하여 고민하여 여러 방안을 생각 했습니다.
1. private (set) var name
2. getName() -> String
3. super.description.componets ~~~~~~ 
4. closure

 위와 같은 방법 중 2,3 은 옳지 않은 방법이며, 4로는 super.name에 접근할 수 없었기에 1을 택했고, 질문에 대한 답변으로도 긍정정인 면이 있었기에 1 의 방법을 택했습니다.

## Set
isHot()과 같은 method에서 반환 값을 Dictionary<Drink,Int> 로 구현했을때 각 Drink마다 manufacturing이 상이하므로, 중복된 상품의 이름이 나왔습니다. 이를 해결하기 위해, 반환값을 Set<Drink.name, Int>로 변경했습니다.

## DownCasting
isHot()과 같은 method의 경우 매개변수로 Drink로 받는데, 하위 class중 isHot이 없는 method를 구분해야 했습니다. protocol Hotable을 만들어 매개변수로 protocol을 지정하는 것도 생각해봤지만, class 내부 property의 Access Control이 불가하므로, 이를 제외하였습니다. 이 외의 방법으로 DownCasting을 생각해내어 적용했습니다.

dahun-lee-daji committed:
convertToString() 리턴값 수정

string(for :) -> String?에서
(from: ) -> String 으로 변경

dahun-lee-daji committed:
chocoMilk : Milk 추가, Milk에 유통기한 만료 여부 기능 구현

Date :
method 추가 : 오늘로부터 1~7일 전 Date를 리턴  -> 제조일자 생성을 위함

Drink :
method 추가 :
제조일로 부터 n일 뒤 Date  반환
현재 시간과 매개변수 :Date를 비교하는 method

Milk :
속성 추가 : expirationLimit
method 추가 : 제조일자와 유통기한을 비교하여 출력

chocoMilk :
속성 추가 : realCholoate
method 추가 : realCholate를 반환

dahun-lee-daji committed:
Add StrawberryMilk : Milk

Straw~ Milk 's new property : cochineal

cochineal is color ingredient that looks like red, come from bug

method : isCochineal() -> Bool

dahun-lee-daji committed:
Add property to Class Soda

New property = package
New method = packingmaterial()

new enum = Materials

dahun-lee-daji committed:
Add class Cola : Soda

Modify Soda.init
New property in Cola = unsweetened
New method = isUnsweetened()

dahun-lee-daji committed:
Add class Cider : Soda

New property in Cider = calories
New method = productCalorie() -> String

dahun-lee-daji committed:
Modify class Coffee

Add property : caffeine : Double?
Add method : productCaffeine() -> String
if caffeine is nil, return "deCaffeine"

dahun-lee-daji committed:
Add class TOP : Coffee

New property in TOP = hot
New method = isHot()

dahun-lee-daji committed:
Add class Cantata : Coffee

New property = Beans
New method = productBeans()
New enum Beans

dahun-lee-daji committed:
3단계 class에 convenience init 추가, testCode 추가

test를 위한 convenience init 추가.
test 코드 작성

dahun-lee-daji committed:
폴더 정리 : enum, struct VendingMachine추가

dahun-lee-daji committed:
자판기 구현

자판기 금액 올리는 method
특정 상품 재고 추가 method
구매 가능한 재고 목록 반환 method -> 수정 필요
음료수 구매 method
잔액 확인 method

dahun-lee-daji committed:
구매 기능 로직 수정, 전체 재고 반환 method 구현

buyProduct() : 구매한 상품 목록에 넣는 내용 추가.
이를 위해 Drinks.remove의 로직 및 반환값 수정

showStock() : 현재재고를 dictionary로 반환
Drink hashable채택,

dahun-lee-daji committed:
Drink : Hashable 수정

프로토콜 채택 내용 수정

dahun-lee-daji committed:
요구 사항 구현

유통기한 지난 재고 리턴 method
뜨거운 음료 return method
구매목록 return method

dahun-lee-daji committed:
Bug Fix, testCode 작성

availableList() : 제품명이 중복되어 나오 던 것을 Set으로 변경하여 해결

buyProduct() : 제품의 Type을 비교하여 해결.-> Drinks의 매개변수 변경 -> Drink.isSameName을 isSameType으로 변경

showStock() : 제품의 이름과 수량만을 표시하도록 변경

hotDrinks() : 제품의 이름만을 표시하도록 변경

Drink : name을 접근 가능하도록 변경

godrm commented on VendingMachineApp/VendingMachineApp/Model/ChocoMilk.swift:
```diff
@@ -0,0 +1,27 @@
+
+import Foundation
+
+//http://emart.ssg.com/item/itemView.ssg?itemId=0000007095603&siteNo=6001&salestrNo=2034
```
이건 그냥 본인이 참고한 쇼핑 아이템이니까 소스에 포함할 필요는 없지 않을까요?
소스 코드를 이해하는 데 꼭 필요한 정보일까요?

dahun-lee-daji replied:
불필요한 정보입니다. 추후 제가 해당 상품 내용을 보려고 적어놓은 내용입니다ㅎㅎ; 삭제하고 진행하겠습니다!

godrm commented on VendingMachineApp/VendingMachineApp/Model/ChocoMilk.swift:
```diff
@@ -0,0 +1,27 @@
+//http://emart.ssg.com/item/itemView.ssg?itemId=0000007095603&siteNo=6001&salestrNo=2034
+class ChocoMilk: Milk {
+    
+    private let realChocolate : Bool //how many day after manufacturingDate
```
두 가지 측면이 있습니다.
- 주석과 코드가 불일치합니다. 이런 주석은 정말 혼란을 줄 수 있습니다. 없는게 낫죠
- realChocolate 함유가 된건지 여부만 Bool로 지정할 수도 있지만, 이런 경우는 함유량을 모델링하고 일정 기준이 넘어가면 true/false로 주는 메소드를 제공하는 게 더 좋은 방법입니다. 속성을 선언할 때도 고려해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Coffee.swift:
```diff
@@ -0,0 +1,31 @@
+
+class Coffee: Drink {
+    
+    // 카페인 추가 -> optional 선언으로 decaffein 구현
+    // 하위 : 원두 품종
+    //  뜨거움 여부
+    
+    private let caffeine : Double?
```
정보를 표현하는 주석이 구체적이지 않고 혼란을 줄 수 있으면 차라리 없는게 낫습니다. 참고하세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Coffee.swift:
```diff
@@ -0,0 +1,31 @@
+            return "deCaffeine"
+        }
+        
+        return "\(nonOptional) mg"
```
데이터를 Double로 갖고 있는데 판단해주는 것이 String인 것은 적합하지 않습니다. 
용량을 가져가서 그걸로 판단해서 어떻게 출력할 지는 뷰와 관련이 높습니다. 데이터를 보관하는 객체가 판단할 필요는 없습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Cola.swift:
```diff
@@ -0,0 +1,26 @@
+        self.init(brand: "CoCa-Cola", volume: 215, charge: 574, name: "코카콜라제로", manufacturing: Date.init(), package : Materials.steel, unsweetened : true) //제조일자 : now
+    }
+    
+    func isUnsweetened() -> Bool {
+        return unsweetened
+    }
```
이런 경우도 데이터 선언과 메소드를 분리해서 Bool 타입으로 저장하지 않고 당분 함유량 같은 값을 보관하는 형태가 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Drink.swift:
```diff
@@ -7,22 +7,72 @@
-class Drink: CustomStringConvertible {
+class Drink: CustomStringConvertible, Hashable {
+    
+    func hash(into hasher: inout Hasher) {
+        hasher.combine(self.brand)
+        hasher.combine(self.charge)
+        hasher.combine(self.manufacturing)
+        hasher.combine(self.name)
+        hasher.combine(self.volume)
+    }
+    
+    var hashValue: Int {
+        var hasher = Hasher()
+        self.hash(into: &hasher)
+        return hasher.finalize()
+    }
```
hash(into:)와 hashValue 차이점은 뭔가요? 꼭 둘 다 선언해야 하던가요?

godrm commented on VendingMachineApp/VendingMachineApp/Model/Drink.swift:
```diff
@@ -7,22 +7,72 @@
+    }
+    
+    func afterManufacuring(days : Int) -> Date {
+        return Date.init(timeInterval: TimeInterval(days * 3600 * 24), since: manufacturing)
```
이 메소드는 음료 속성과 관련이 없어보이네요?
만약 기준 제조일부터 며칠 지난건지 판단하려는 거라면 Date와 관련이 있지 않을까요?

godrm commented on VendingMachineApp/VendingMachineApp/Model/Drink.swift:
```diff
@@ -7,22 +7,72 @@
+    }
+    
+    /// 지금과, date를 second까지 비교, date가 더 미래라면 true, else = false. 낮은 경우의 수로 지금과 date가 동일 시간이어도, false로 처리함.
+    func compareNowWith(date : Date) -> Bool {
+        var bool = false
+        if Date.init().compare(date) == .orderedAscending {
+            bool = true
+        }
+        
+        return bool
+    }
```
이 메소드도 내부 속성을 안 쓰는것 같은데요? 어떤 타입과 관련이 있는걸까요?

godrm commented on VendingMachineApp/VendingMachineApp/Model/Drinks.swift:
```diff
@@ -0,0 +1,42 @@
+            $0.isSameType(with: productType)
+        })
+        else {
+            throw errorCase.outOfIndex
```
에러를 잘 정의하셨네요. 이런 경우 그냥 옵셔널 정도로 처리할 지 throw 할 지 결정이 중요합니다.
에러는 옵셔널로 의미를 표현하기 어렵거나 2가지 이상 다양한 경우에, 그리고 정상적으로 처리하기 어려운 심각한 상황에서 쓰도록 권장하는 편입니다. 이정도는 옵셔널도 안전하게 처리할 수 있을 것 같습니다.

dahun-lee-daji replied:
Drinks.remove()는 매개변수 productType과 같은 타입을 찾아서 pop합니다.
때문에, firstIndex를 찾지 못했을 경우, 옵셔널 처리를 할 경우, drinks.remove가 틀린 동작을 할 수 있다고 여겨, throw 했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Drinks.swift:
```diff
@@ -0,0 +1,42 @@
+    }
+    
+    func doClosure(closure : (([Drink]) -> Void)) {
+        closure(drinks)
```
클로저를 넘기는 방법도 다양한 방법을 시도해보세요. 
처음에는 이렇게 다 넘기는 것도 가능하지만
조건에 따라서 해당 음료수만 골라서 넘기는 메소드를 따로 구현하는 것도 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Milk.swift:
```diff
@@ -7,9 +7,22 @@
-    init() {
-        super.init(brand: "서울우유", volume: 200, charge: 830, name: "초코우유", manufacturingDate: Date.init()) //제조일자 : now
+    
+    private let expirationLimit : Int //how many day after manufacturingDate
```
날짜/기간과 관련된 것은 Int 보다는 Date 혹은 TimeInterval 같은 타입도 고려해보세요.

dahun-lee-daji replied:
해당 property를 추가 할 때, "몇 일"을 모델링 하려고 했습니다.
실제 method 구현 시에도, 3일 뒤라고 했을 때, 이를 매개변수로 제공하여 TimeInterval : expirationLimit * 3600 * 24 를 이용하거나 Calendar를 사용할 때에도 Int 가 더 유리하다고 생각했습니다.

expirationLimit을 Date로 사용 할 경우, 몇 일 뒤를 init 할 경우, 3일 뒤를 표현하기 어려울 것이라 생각했으며,
TimeInterval을 사용 할 경우, second를 init의 매개변수로 제공해야 하는데, 이는 사람이 보기 쉽지 않으며 Data를 가공하기 어려울 것이라 생각했습니다.

godrm replied:
그렇지 않습니다. Date 생성자나 메소드 중에도 3일뒤를 계산할 수 있는 메소드가 있습니다.
더구나 3일뒤 라는 것을 직접 계산하는 게 절대 수월하지 않습니다. 섬머타임이 있다거나 타임존이 바뀌거나, 비행기에서 날짜가 바뀌는 경우는 더 계산이 복잡해집니다. Date는 그런 것을 모두 고려해서 구현되어 있는 타입니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Milk.swift:
```diff
@@ -7,9 +7,22 @@
+    }
+    
+    /// 지금과, date를 second까지 비교, date가 더 미래라면 true, else = false. 낮은 경우의 수로 지금과 date가 동일 시간이어도, false로 처리함.
+    func validate() -> Bool {
```
validate 라는 함수를 주석을 설명해주는 시도는 좋습니다.
다만 함수 이름으로 그것을 더 잘 표현하는 게 주석 보다 좋습니다. 
validate가 구체적으로 뭘한다는 것을 알 수 없는데, 위에 주석보다 명확한 함수 이름이라면 주석을 뺄 수 있을 것 같습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Soda.swift:
```diff
@@ -9,7 +9,15 @@ import Foundation
-    init() {
-        super.init(brand: "CoCa-Cola", volume: 215, charge: 574, name: "코카콜라제로", manufacturingDate: Date.init()) //제조일자 : now
+    
+    private let package : Materials
```
Materials 타입을 Soda에서만 쓰는 거라면 여기 근처에... 아니면 다른 파일에 있어도 될 것 같습니다.

dahun-lee-daji replied:
넵 그래서 enum/Materials파일에 있습니다!

godrm commented on VendingMachineApp/VendingMachineApp/Model/StrawberryMilk.swift:
```diff
@@ -0,0 +1,26 @@
+
+class StrawberryMilk: Milk {
+    
+    private let cochineal : Bool //how many day after manufacturingDate
```
이 속성도 정보가 주석과 불일치하네요.
색소 정보인가요?

dahun-lee-daji replied:
주석 제거하였습니다.
해당 내용은 색소 정보가 맞습니다!

godrm commented on VendingMachineApp/VendingMachineApp/Model/TOP.swift:
```diff
@@ -0,0 +1,27 @@
+//http://emart.ssg.com/item/itemView.ssg?itemId=0000007323904&siteNo=6001&salestrNo=2034&tlidSrchWd=TOP&srchPgNo=1&src_area=elist
+class TOP: Coffee {
+    
+    private let hot : Bool
```
뜨겁다는 정보도 Bool 보다는 온도를 저장하고 기준 온도보다 낮냐 높냐를 판단해주는 메소드가 있는게 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -0,0 +1,113 @@
+import Foundation
+
+struct VendingMachine {
+    private var money : Int
```
money도 타입으로 선언해보면 어떨까요?

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -0,0 +1,113 @@
+                else {
+                    returnDic.updateValue(returnDic[$0.name]! + 1, forKey: $0.name)
+                }
+            })
```
이렇게 두 단계로 클로저를 두는 것은 그냥 배열을 가져온 것과 사실 다르지 않습니다. 
클로저를 사용할 때도 구체적인 동작을 고려해보세요.

dahun-lee-daji replied:
배열을 그냥 가져온 것 과 다르지 않다는 건 .forEach()를 말씀하신 건가요?
구체적인 동작을 고려해보라는 말씀이 어떤 것인지 잘 모르겠습니다..ㅠㅠ

godrm replied:
클로저를 활용하라는 측면은 값을 가져오지 않고 로직을 전달하기 위한 방법 중에 하나입니다.
더구나 내부 속성이 2차원 배열인데 doClosure와 drinks.forEach로 2차원 배열을 그대로 탐색하는 방법은 drinks 라는 내부 배열을 그대로 가져간 것과 별로 다르지 않다는거죠. 
클로저를 활용하지 않고 이런 메소드가 `stock.show() -> Dictionary<String,Int>`있는 것보다 오히려 더 안좋다는 의미입니다. 모든 메소드를 클로저를 넘기는 방식으로 구현하지 않아도 됩니다. 
클로저로 내부에서 구현해도 될 것 까지 통째로 클로저로 제공하라는 게 아닙니다.

dahun-lee-daji replied:
아 제가 클로저로 넘기는 거에 너무 꽂혀있던 것 같습니다.
답변 감사합니다!

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -0,0 +1,113 @@
+                        returnDrinkArr.append(downCasting)
+                    }
+                }
+            })
```
모든 것을 하나의 클로저로 처리해야 하는 건 아닙니다.

dahun-lee-daji replied:
closure에 대한 피드백이 이해가 조금 어렵습니다 ㅜ
doClosure()를 반복적으로 사용하지 말라는 말씀이신가요?  -> doClosure()외에 특정 음료에 대한 closure method를 추가를 말씀하시는 걸까요?

godrm replied:
모든 것을 클로저로 구현하지 말라는 거죠. 
클로저가 아니라 정해진 형식대로 메소드를 호출하고 return 값을 전달해도 되는 것을 모두다 클로저로 구현할 필요는 없습니다.

godrm reviewed (changes requested):
수고하셨습니다. 다음 피드백 내용을 확인해서 개선해보세요.

dahun-lee-daji committed:
불필요한, 틀린 주석 제거. Coffee.productCaffeine() 반환 값 수정

불필요한 주석 : convenience init에 쓰인 상품에 대한 정보 링크

틀린 주석 : 실수로 들어간 내용

Coffee.productCaffeine() -> Double로 변경.
값을 반환하고, 해당 내용의 표기는 상위 모듈에서 조절하도록 함.

dahun-lee-daji committed:
초콜렛 여부를 함량으로 판단하도록 수정

기존 bool -> 함유량 Double
20% 이상일 경우, 초콜렛으로 판별. -> 한국 법 기준

dahun-lee-daji committed:
무가당 여부 판별 로직 수정

존 bool -> 함유량 Double
가당 함유량 2% 미만일 경우 무가당으로 판별함.

dahun-lee-daji committed:
Hasble 프로토콜 채택 내용 수정

Swift5 기준, hashValue 직접 설정 : deprecated

dahun-lee-daji committed:
method 소속 변경

Drink -> Date
compareNowWith()
afterDays(days : Int) -> Date
해당 이동을 위해 매개변수 및 property의 접근지정자를 수정

dahun-lee-daji committed:
method명 변경

validate -> isWithInExpiration

dahun-lee-daji committed:
주석 삭제

dahun-lee-daji committed:
Modify TOP`s property : hot -> temperature

뜨거운 여부 -> 온도
70도 이상일 경우 뜨거운 음료로 판별하도록 로직 변경.

dahun-lee-daji committed:
money : Int에서 SafeBox로 변경

SafeBox 타입 추가.
기능 : 입금, 출금, 잔액 확인
money 관련 메소드 수정
money가 주체가 되어 돈을 관리하므로, Drink의 payfor를 음료 가격 반환으로 변경

dahun-lee-daji commented:
1차 수정 내용 : 
불필요, 틀린 내용의 주석 제거.
Coffee.productCaffeine() -> String에서 Double로 수정.
TOP의 hot : Bool, ChocoMilk의 realChocolate: Bool, Cola의 unsweetened: Bool 을 구체적 수치로 변경, 및 관련 메소드 수정
유통기한 확인 method = validate() 메소드 명 변경
Drink 내부의 Date관련 method 로직 및 매개변수 수정하여 Date.Extension으로 이동.
money : Int 에서 SafeBox로 변경.

Hashable프로토콜 채택 내용 수정 : `var hashValue: Int {}` 내용 삭제.
-> swift5에서 변경된 내용. Hasher가 hashValue를 생성해주기 때문에, func hash 내에서 hasher.combine method를 사용하여 essential parts를 mixing해주기만 하면 됨.

반영되지 않은 피드백 : closer method 관련 내용
-> 피드백 내용의 이해가 어렵습니다 ㅜㅜ
closure에 대한 피드백이 이해가 조금 어렵습니다 ㅜ

doClosure()를 반복적으로 사용하지 말라는 말씀이신가요? -> doClosure()외에 특정 음료에 대한 closure method를 추가를 말씀하시는 걸까요?

Dictionary를 만들어 재고를 보여주는 method에서 배열을 가져온 것과 다름없다고 말씀하신 것과, 구체적인 동작을 고려해보라는 말씀이 이해가 잘 되지않습니다.
drink.name을 key로 하여, Dictionary내에 해당 key에 대한 value 가 없을 경우, 값을 넣어주며, 있을경우 value += 1  하는 동작입니다. 
-> drinks.forEach를 사용한다면 배열을 가져온 것과 같은 내용인가요?, 제 생각에는 구체적인 동작을 하고 있는 것 같은데, JK의 의도가 뭔지 파악이 어렵습니다 ㅜㅜ

godrm commented on VendingMachineApp/VendingMachineApp/Model/ChocoMilk.swift:
```diff
@@ -7,21 +7,20 @@
     
     func isRealCholate() -> Bool {
-        return realChocolate
+        return chocolateContent > 20
```
여기서 기준값 20을 상수로 선언하거나 사위 타입으로 선언하고, 매개변수로 전달하는 구조도 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/ChocoMilk.swift:
```diff
@@ -7,21 +7,20 @@
 class ChocoMilk: Milk {
     
-    private let realChocolate : Bool //how many day after manufacturingDate
+    private let chocolateContent : Double // percent
```
비율보다는 용량이 더 좋습니다. 비율은 생각보다 정확한 값이 아닙니다. 오차가 생길 수 있습니다.
내용물로 비율을 계산하는 게 더 정확합니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Cola.swift:
```diff
@@ -9,18 +9,18 @@ import Foundation
 class Cola: Soda {
     
-    private let unsweetened : Bool
+    private let sugarContent : Double
```
Double 이나 Float를 사용할 때는 오차 범위나 표현범위를 고려해야 합니다. 참고하세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Drink.swift:
```diff
@@ -40,7 +33,7 @@ class Drink: CustomStringConvertible, Hashable {
     private let charge: Int
     private (set) var name: String
-    private let manufacturing: Date
+    private (set) var manufacturing: Date
```
값을 가져가기 보다는 메소드를 제공하려고 노력해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/extention/Date.swift:
```diff
@@ -18,4 +18,18 @@ extension Date {
+    }
+    
+    func afterDays(days : Int) -> Date {
+        return Date.init(timeInterval: TimeInterval(days * 3600 * 24), since: self)
```
add() advanced() 같은 메소드와 어떻게 다를까요?

godrm commented:
고민이 드러나는 것 같습니다. 다만 제가 더 구체적으로 이렇게 바꾸세요. 라고 하지 않는 것은 힌트정도로 답을 찾아가길 바래서입니다.
제가 의도한 데로 고치는 게 중요한 게 아니라 스스로 무엇을 고쳐야 할 지 깨닫는 게 더 중요합니다. 
제 피드백에 맞춰서 그걸 해결하려고 하지마시고, 이해가 안되면 더 질문을 댓글로 남기세요. 고민하다가 수정하지 못하는 것보다 질문하는 게 더 좋습니다. 리뷰는 저의 의견에 맞춰서 일방적으로 구현하자는 게 아닙니다. 코드로 저를 설득하려고 하셔야 합니다.
